### PR TITLE
Add wildcarding for searching field names

### LINKED ISSUE #406: As a user, I want to query for documents where a specific search field exists in the document
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Data User

### 💪 Motivation

...so that I can find products that contain (or do not contain) a specific field value in the document.

### 📖 Additional Details

The `exists` operator allows users to query for documents based on the presence (or absence) of specific fields in the document.

> **⚠️ Syntax Update:** The original implementation used **postfix** notation (e.g., `field exists`). Based on API Working Group feedback, the syntax has been updated to **prefix** notation (e.g., `exists field`) in #727 for more intuitive usage.

---

#### Original Implementation (Postfix - Deprecated)

**Syntax:**
- Check if exact field exists: `pds:Target/pds:name exists`
- Check if field does NOT exist: `not (pds:Investigation/pds:stop_date_time exists)`
- Regex pattern matching: `"pds:Target.*" exists` - matches any field starting with `pds:Target`
- Complex regex patterns: `".*Bounding_Coordinates.*" exists`

---

#### Updated Syntax (Prefix - See #727)

**Syntax:**
- Check if exact field exists: `(exists pds:Target/pds:name)`
- Check if field does NOT exist: `(not exists pds:Investigation/pds:stop_date_time)`
- Regex pattern matching: `(exists "pds:Target.*")` - matches any field starting with `pds:Target`
- Complex regex patterns: `(exists ".*Bounding_Coordinates.*")`

---

**Implementation Notes:**
- For exact field names (unquoted), the query checks for that specific field in the document
- For quoted strings, the value is treated as a Java regex pattern and matched against all known field names from the OpenSearch mapping
- When using regex patterns, all matching field names are retrieved and an existence check is created for each
- Returns error if regex pattern matches no known fields
- `exists` returns true only if the field is present AND has a non-null/non-empty value
- For multi-valued fields, `exists` returns true if at least one value is present

**Use Case Examples (Updated Syntax):**
* Find all products that have bounding box coordinates: `(exists cart:Bounding_Coordinates/cart:north_bounding_coordinate)`
* Find investigations that are still ongoing (no stop date): `(not exists pds:Investigation/pds:stop_date_time)`
* Find products with any target field using regex: `(exists "pds:Target.*")`
* Find products missing author information: `(not exists pds:Citation_Information/pds:author_list)`
* Complex pattern matching: `(exists ".*Coordinates.*")` finds all fields containing "Coordinates"

Related: See #402 for related query operator issues.
Related: See #727 for the syntax update from postfix to prefix notation.

### Acceptance Criteria

**Given** I am querying the registry API
**When** I use `(exists field_name)` with an exact field name in my query
**Then** the API returns only documents where the specified field exists and has a non-null/non-empty value

**Given** I am querying the registry API
**When** I use `(not exists field_name)` in my query
**Then** the API returns only documents where the specified field does not exist or is null/empty

**Given** I am querying for a multi-valued field
**When** I use `(exists field_name)` and at least one value is present
**Then** the API returns that document in the results

**Given** I want to check for fields matching a regex pattern
**When** I use `(exists "regex_pattern")` with a quoted string
**Then** the API:
- Retrieves all field names from the OpenSearch mapping
- Matches the regex against all known field names
- Returns documents where any matched field exists with a non-null/non-empty value
- Returns an error if the regex matches no known field names

**Given** I combine `exists` with other query operators
**When** I use queries like `(exists pds:Target/pds:name) and pds:Target/pds:type eq "Planet"`
**Then** the API correctly applies both conditions and returns matching documents

### ⚙️ Engineering Details

Implementation completed in PR #700 (postfix syntax):
- Updated ANTLR4 lexer grammar (`Search.g4`) to support postfix `EXISTS` keyword with both FIELD and STRINGVAL tokens
- Updated `Antlr4SearchListener` to handle existence checks:
  - Exact field name matching for unquoted field names
  - Java regex pattern matching for quoted strings
  - Retrieves mapping from OpenSearch via `ProductsController.productPropertiesList()` for regex matching
  - Generates OpenSearch `ExistsQuery` for each matched field
- Proper handling of NOT for negation
- Error handling when regex matches no fields

**Syntax Update:** See #727 for the update from postfix to prefix notation based on API Working Group feedback.

### LINKED ISSUE #611: As a user, I want to search by a full/unique hierarchical path for a specific attribute
### Checked for duplicates

No - I haven't checked

### 🧑‍🔬 User Persona(s)

Data User, Instrument Scientist, Rover Operations Team, Navigation Specialist

### 💪 Motivation

The current PDS Search API implementation flattens searchable metadata to `ns:Class.ns:attribute`, which eliminates the ability to search for attributes where a 3rd or 4th level of the PDS4 hierarchy matters in the definition of that value. Multiple discipline nodes have expressed a need to support queries where greater hierarchical context is necessary, particularly for:

1. **Nested array structures** - Where the same attribute appears in multiple contexts
2. **Context-dependent attributes** - Where attribute values depend on their position in the XML hierarchy
3. **Complex metadata queries** - Where filtering requires matching attributes within the proper nested structure

This requirement focuses on implementing **simple hierarchical path searches using glob patterns** to navigate nested structures.

### 📖 Additional Details

**Problem Statement:**
Current flattening approach: `ns:Class.ns:attribute`
Needed capability: Support for hierarchical paths with wildcards:
- `*.ns:Class.ns:attribute` - Match any parent
- `*.ns:Class.ns:Class2.ns:Class3.ns:attribute` - Match specific nested path
- `*.ns:Class.*` - Match any attribute within a class

**Primary Use Cases:**

**US1: Query Nested Attributes by Full Path**
As a data user, I want to query attributes using their full hierarchical path (e.g., `*.img:Imaging.img:Exposure.img:exposure_duration < 0.001`), so I can filter products based on deeply nested attribute values.

**US2: Wildcard Path Matching**
As an instrument scientist, I want to use wildcards in hierarchical paths (e.g., `*.geom:Device_Temperature_Index.geom:index_value_temperature > 30`), so I can query nested attributes without knowing the complete parent hierarchy.

**US3: Hierarchical Context Filtering**
As a rover operations team member, I want to query attributes within specific nested contexts (e.g., `*.geom:Motion_Counter.geom:Motion_Counter_Index.geom:index_value_number == 33`), so I can filter based on hierarchical relationships.

**Proposed Solution:**
- Support hierarchical path syntax using dot notation: `ns:ClassRoot:ns:Class.ns:Class2.ns:attribute`
- Support wildcards in paths: `*` to match any class/attribute at that level
- Maintain backwards compatibility for existing flattened queries (e.g., `ns:Class.ns:attribute eq "foobar"`)
- Leverage OpenSearch nested field types for hierarchical queries
- Translation layer to convert hierarchical path queries to OpenSearch nested queries

**Example Query Syntax:**

```
// Full hierarchical path
img:Imaging.img:Exposure.img:exposure_duration < 0.001

// Wildcard parent
*.img:Exposure.img:exposure_duration < 0.001

// Wildcard attribute
img:Imaging.img:Exposure.* eq "some_value"

// Multiple levels with wildcards
*.geom:Device_Temperature_Index.geom:index_value_temperature > 30
```

**OpenSearch Implementation:**
These queries will translate to OpenSearch nested queries with proper path scoping using dot notation to navigate the nested field hierarchy.

### Acceptance Criteria

**Given** a PDS4 product with nested metadata structures
**When I perform** a query using hierarchical dot notation with optional wildcards
**Then I expect** the API to correctly filter based on nested attribute values within their proper hierarchical context

**Given** a query with wildcard path segments
**When I execute** the query (e.g., `*.ns:Class.ns:attribute eq "value"`)
**Then I expect** the API to match all products where that attribute exists at any level in the hierarchy with the specified value

### ⚙️ Engineering Details

Implementation considerations:
- Update lexer/parser to support hierarchical path syntax with dots and wildcards
- Implement path-to-OpenSearch-nested-query translator
- Ensure proper handling of namespace prefixes in nested paths
- Add validation for path syntax
- Maintain backwards compatibility with current flattened attribute queries

### 🎉 I&T

Test coverage should include:
- Simple nested paths (2-3 levels deep)
- Paths with wildcards at various positions
- Mixed queries combining hierarchical and flattened paths
- Edge cases: invalid paths, non-existent paths, deeply nested structures

### LINKED ISSUE #622: Add Search Functionality to Find When Class / Attribute Has Been Used / Exists
### Are you sure this is not a new requirement or bug?

Yes

### Task Type

Theme

### 💡 Description

Similar to SQL `exists()` functionality, we want the API to be able to support this.

### Engineering details

#647 is an engineering requirement of the ticket.

### LINKED ISSUE #712: As a user, I want the exists operator to match OpenSearch's native behavior for all fields
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Data User

### 💪 Motivation

...so that the `exists` operator behaves consistently with OpenSearch's native functionality, providing predictable and intuitive results when querying for fields regardless of whether they exist in the schema.

---

**NOTE: The exact implementation of this requirement will impacted by:**
- https://github.com/NASA-PDS/registry-api/issues/727

---

### 📖 Additional Details

This requirement extends the `exists` operator implementation from #406 to ensure the API behavior matches OpenSearch's native behavior for all field existence queries.

**Context:**
Following the completion of #406 in PR #700, additional behavior clarifications were identified in the PR review discussions (https://github.com/NASA-PDS/registry-api/pull/700) regarding how to handle fields that don't exist in the registry schema.

**Problem:**
The API should provide consistent, predictable behavior that matches OpenSearch's native functionality. Users familiar with OpenSearch expect certain behavior when querying for field existence, and the API should honor those expectations.

**Proposed Behavior:**

The API should match OpenSearch's native behavior for all `exists` queries:

1. **Field exists in schema with matching documents**:
   ```
   pds:Internal_Reference.reference_type exists
   ```
   - Returns only documents where this field has a value
   - Standard expected behavior

2. **Field exists in schema with no matching documents**:
   ```
   pds:Some_Real_Field_With_No_Values exists
   ```
   - Returns 0 results (no documents have this field populated)
   - Consistent with OpenSearch behavior

3. **Field does NOT exist in schema**:
   ```
   pds:Nonexistent_Field exists
   ```
   - Returns 400 error with a clear, descriptive message identifying the unknown field name

4. **NOT exists with field in schema**:
   ```
   not (pds:Internal_Reference.reference_type exists)
   ```
   - Returns documents where this field does NOT have a value
   - Standard expected behavior

5. **NOT exists with field NOT in schema**:
   ```
   not (pds:Nonexistent_Field exists)
   ```
   - Returns 400 error with a clear, descriptive message identifying the unknown field name

6. **Regex pattern matching fields**:
   ```
   "pds:Internal_Reference.*" exists
   ```
   - Returns documents where ANY matching field has a value
   - Standard expected behavior

### Acceptance Criteria

**Given** a user queries for an exact field name that exists in the OpenSearch mapping with documents containing values
**When** the user uses the `exists` operator (e.g., `pds:Internal_Reference.reference_type exists`)
**Then** the API returns only documents where this field has a value

**Given** a user queries for an exact field name that exists in the OpenSearch mapping but no documents have values for it
**When** the user uses the `exists` operator (e.g., `pds:Some_Real_Field_With_No_Values exists`)
**Then** the API returns 0 results

**Given** a user queries for an exact field name that does NOT exist in the OpenSearch mapping
**When** the user uses the `exists` operator (e.g., `pds:Nonexistent_Field exists`)
**Then** the API returns a 400 error with a clear, descriptive message identifying the unknown field name (e.g., "Field 'pds:Nonexistent_Field' does not exist in the registry schema")

**Given** a user queries for an exact field name that exists in the OpenSearch mapping
**When** the user uses the `not exists` operator (e.g., `not (pds:Internal_Reference.reference_type exists)`)
**Then** the API returns only documents where this field does NOT have a value

**Given** a user queries for an exact field name that does NOT exist in the OpenSearch mapping
**When** the user uses the `not exists` operator (e.g., `not (pds:Nonexistent_Field exists)`)
**Then** the API returns a 400 error with a clear, descriptive message identifying the unknown field name (e.g., "Field 'pds:Nonexistent_Field' does not exist in the registry schema")

**Given** a user queries with a regex pattern that matches one or more fields
**When** the user uses the `exists` operator with a quoted pattern (e.g., `"pds:Internal_Reference.*" exists`)
**Then** the API returns documents where ANY of the matching fields have values

### ⚙️ Engineering Details

_To be filled by engineering team during implementation_

Considerations:
- Ensure the implementation delegates field existence logic to OpenSearch rather than pre-validating against the mapping
- Test with fields that exist vs. don't exist in the schema
- Test with fields that exist in schema but have no values in any documents
- Verify `not exists` behavior for both existent and non-existent fields
- Ensure regex patterns work correctly

### 🎉 I&T

_To be filled by engineering team during implementation_

Integration tests should verify:
- Queries for existent fields with values return correct results
- Queries for existent fields with no values return 0 results
- Queries for non-existent fields (exists or not exists) return 400 with a descriptive error message identifying the unknown field
- Regex patterns matching multiple fields return documents with ANY matching field

### PR DESCRIPTION
**Pull request by al-niessner**
## 🗒️ Summary
Allow simple like globbing for field names.

## ⚙️ Test Data and/or Report
Local test below should pass.

## ♻️ Related Issues
Closes #406
Closes #611 
Closes #622 
Closes #712 

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Documentation
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.

### Maintenance
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue or Jira Ticket.
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

### CODE REVIEW

**Comment by al-niessner** [comment]
@tloubrieu-jpl 

I see that all of the PRs are showing broken. Let me know when I should expect the tests to work and I will make corrections to the tests. I removed using regex as an exist wildcard option and just use glob everywhere equally. Hence, I will need to fix that test when they are working again.

**Comment by jordanpadams** [comment]
@tloubrieu-jpl I will leave this for you to test because I lost track of the failing integration tests.

File: service/src/main/java/gov/nasa/pds/api/registry/model/Antlr4SearchListener.java [hunk 1]
```diff
@@ -37,9 +37,11 @@ enum operation {
 
   private static final Logger log = LoggerFactory.getLogger(Antlr4SearchListener.class);
 
+  private boolean isAnyWildcard = true;
   private BoolQuery.Builder queryBuilder = new BoolQuery.Builder();
   private conjunctions conjunction = conjunctions.AND; // DEFAULT
 
+  private final ArrayList<String> fieldnames = new ArrayList<String>();
```
Comment (tloubrieu-jpl) [inline service/src/main/java/gov/nasa/pds/api/registry/model/Antlr4SearchListener.java]: Shound `fieldnames` be `fieldNames` with N upper case.

File: service/src/main/java/gov/nasa/pds/api/registry/model/Antlr4SearchListener.java [hunk 2]
```diff
@@ -110,18 +153,14 @@ public void exitOrStatement(SearchParser.OrStatementContext ctx) {
...
 
-  @Override
-  public void enterComparison(SearchParser.ComparisonContext ctx) {}
-
   @Override
   public void exitComparison(SearchParser.ComparisonContext ctx) {
     log.debug("Exit comparison");
-    final String left = SearchUtil.jsonPropertyToOpenProperty(ctx.FIELD().getSymbol().getText());
 
+    BoolQuery.Builder wild = new BoolQuery.Builder();
```
Comment (tloubrieu-jpl) [inline service/src/main/java/gov/nasa/pds/api/registry/model/Antlr4SearchListener.java:160]: The OpenSearch logic should be implemented in the object https://github.com/NASA-PDS/registry-api/blob/develop/service/src/main/java/gov/nasa/pds/api/registry/search/RegistrySearchRequestBuilder.java , in case one day we change the OpenSearch backend of connection client.

We could have a function `getFieldNamesFromRegex` or something similar... I might misunderstand what is done here, but have the a method in the RegistrySearchRequestBuilder will help to formlize what is the interface with OpenSearch and keep all the OpenSearch related code in a single place.

**Comment by tloubrieu-jpl** [comment]
@al-niessner I fixed the integrated test  in the registry repository, in the `develop` branch. Many tests are failing with this PR, for example:
{{baseUrl}}/products/urn:nasa:pds:mars2020.spice:spice_kernels:mk_m2020/member-of?q=pds:Time_Coordinates.pds:start_date_time eq "2020-07-30T12:51:34Z" does not retunr results anymore.

Some tests which previously worked with the exists operator do not work either.

**Review by tloubrieu-jpl — CHANGES_REQUESTED** [review]
There are 2 change request:

-  code organization: OpenSearch code in the dedicated request builder.
- the integration test need to validate.

File: service/src/main/java/gov/nasa/pds/api/registry/model/Antlr4SearchListener.java (hunk 1 above)
Comment (al-niessner) [reply service/src/main/java/gov/nasa/pds/api/registry/model/Antlr4SearchListener.java]: Ignorance and apathy on my part. I googled it just to see. Fieldnames is a tech word and is commonly used in this situation (DB) frequently. Therefore, all depends on if you prefer fieldnames or field names. Makes absolutely no difference to me.

File: service/src/main/java/gov/nasa/pds/api/registry/model/Antlr4SearchListener.java (hunk 2 above)
Comment (al-niessner) [reply service/src/main/java/gov/nasa/pds/api/registry/model/Antlr4SearchListener.java:160]: I would suggest a ticket for this but leave current code as is. The entire antlr parser is very heavily enmeshed with opensearch. Having 98% enmeshed and 2% not seems odd.

You may need to resign yourself to rewriting the antlr parser for a different back end if that ever comes. If a new back end arrives, then you would have to make Antlr4SearchListener for opensearch and the new back end like SearchRequestBuilder. The antlr stuff is self contained (does not use other parts of the registry-api) and whose whole business is to efficiently map the antlr query language to the back end query language. Adding another layer (SearchRequestBuilder) seems a tad redundant.

**Comment by al-niessner** [comment]
@tloubrieu-jpl 

All postman 4-6 tests pass now on the new PR NASA-PDS/registry#469

**Review by tloubrieu-jpl — APPROVED** [review]
The feature now works, tested with integration.
A new ticket will be creted to have consistent wildcard syntax for properties and values.
